File: daily/Linked List/2058-Find-the-minimum-and-maximum-number-of-nodes-between-critical-points.go
package daily

import "math"

func nodesBetweenCriticalPoints(head *ListNode) []int {
	if head == nil || head.Next == nil || head.Next.Next == nil {
		return []int{-1, -1}
	}

	firstCriticalPoint := -1
	lastCriticalPoint := -1
	minDistance := math.MaxInt32

	prev := head
	curr := head.Next
	currentIndex := 1

	for curr.Next != nil {
		next := curr.Next
		maxima := curr.Val > prev.Val && curr.Val > next.Val
		minima := curr.Val < prev.Val && curr.Val < next.Val

		if maxima || minima {
			if firstCriticalPoint == -1 {
				firstCriticalPoint = currentIndex
			} else {
				dist := currentIndex - lastCriticalPoint
				if dist < minDistance {
					minDistance = dist
				}
			}
			lastCriticalPoint = currentIndex
		}

		prev = curr
		curr = next
		currentIndex++
	}

	if firstCriticalPoint == lastCriticalPoint {
		return []int{-1, -1}
	}

	return []int{minDistance, lastCriticalPoint - firstCriticalPoint}
}

File: daily/Linked List/2058-Find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        if not head or not head.next or not head.next.next:
            return [-1, -1]
        
        first_critical_point, last_critical_point, minDistance = -1, -1, float('inf')
        prev, curr = head, head.next
        current_index = 1
        while curr.next:
            next_node = curr.next
            maxima = curr.val > prev.val and curr.val > next_node.val
            minima = curr.val < prev.val and curr.val < next_node.val
            if maxima or minima:
                if first_critical_point == -1:
                    first_critical_point = current_index
                else:
                    minDistance = min(minDistance, current_index - last_critical_point)
                last_critical_point = current_index
            prev = curr
            curr = next_node
            current_index += 1
        
        if first_critical_point == last_critical_point:
            return [-1, -1]
        return [minDistance, last_critical_point - first_critical_point]